File: src/navigation/calculations.py
import math
from typing import List, Tuple

import numpy as np

EARTH_RADIUS = 6371000


def haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    delta_lat = math.radians(lat2 - lat1)
    delta_lon = math.radians(lon2 - lon1)

    a = math.sin(delta_lat / 2) ** 2 + \
        math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(delta_lon / 2) ** 2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

    return EARTH_RADIUS * c


def bearing_to(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    delta_lon = math.radians(lon2 - lon1)

    x = math.sin(delta_lon) * math.cos(lat2_rad)
    y = math.cos(lat1_rad) * math.sin(lat2_rad) - \
        math.sin(lat1_rad) * math.cos(lat2_rad) * math.cos(delta_lon)

    bearing = math.atan2(x, y)
    return (math.degrees(bearing) + 360) % 360


def cross_track_distance(pos_lat: float, pos_lon: float,
                         wp1_lat: float, wp1_lon: float,
                         wp2_lat: float, wp2_lon: float) -> float:
    d13 = haversine_distance(wp1_lat, wp1_lon, pos_lat, pos_lon) / EARTH_RADIUS
    brng13 = math.radians(bearing_to(wp1_lat, wp1_lon, pos_lat, pos_lon))
    brng12 = math.radians(bearing_to(wp1_lat, wp1_lon, wp2_lat, wp2_lon))

    xtd = math.asin(math.sin(d13) * math.sin(brng13 - brng12))
    return xtd * EARTH_RADIUS


def along_track_distance(pos_lat: float, pos_lon: float,
                         wp1_lat: float, wp1_lon: float,
                         wp2_lat: float, wp2_lon: float) -> float:
    d13 = haversine_distance(wp1_lat, wp1_lon, pos_lat, pos_lon) / EARTH_RADIUS
    xtd = cross_track_distance(pos_lat, pos_lon, wp1_lat, wp1_lon, wp2_lat, wp2_lon) / EARTH_RADIUS

    atd = math.acos(math.cos(d13) / math.cos(xtd))
    return atd * EARTH_RADIUS


def eta_seconds(distance_m: float, groundspeed_ms: float) -> float:
    if groundspeed_ms <= 0:
        return float('inf')
    return distance_m / groundspeed_ms


def meters_to_lat_offset(meters: float) -> float:
    return meters / 111320.0


def meters_to_lon_offset(meters: float, latitude: float) -> float:
    return meters / (111320.0 * math.cos(math.radians(latitude)))


def normalize_heading(heading: float) -> float:
    return heading % 360


def heading_difference(h1: float, h2: float) -> float:
    diff = h2 - h1
    while diff > 180:
        diff -= 360
    while diff < -180:
        diff += 360
    return diff


def project_point(lat: float, lon: float, bearing_deg: float, distance_m: float) -> Tuple[float, float]:
    """Project a point at given bearing and distance from current position.
    Returns (lat, lon) in degrees."""
    d = distance_m / EARTH_RADIUS
    brg = math.radians(bearing_deg)
    lat1 = math.radians(lat)
    lon1 = math.radians(lon)

    lat2 = math.asin(math.sin(lat1) * math.cos(d) + math.cos(lat1) * math.sin(d) * math.cos(brg))
    lon2 = lon1 + math.atan2(math.sin(brg) * math.sin(d) * math.cos(lat1),
                              math.cos(d) - math.sin(lat1) * math.sin(lat2))

    return math.degrees(lat2), math.degrees(lon2)


# ──────────────────── Zone avoidance geometry ────────────────────


def point_in_polygon(lat: float, lon: float, polygon: List[List[float]]) -> bool:
    """Ray casting algorithm to check if point is inside polygon.
    polygon: [[lat, lon], ...] — closed ring (first != last is OK)."""
    n = len(polygon)
    if n < 3:
        return False
    inside = False
    j = n - 1
    for i in range(n):
        yi, xi = polygon[i][0], polygon[i][1]
        yj, xj = polygon[j][0], polygon[j][1]
        if ((yi > lat) != (yj > lat)) and (lon < (xj - xi) * (lat - yi) / (yj - yi) + xi):
            inside = not inside
        j = i
    return inside


def circle_to_polygon(center_lat: float, center_lon: float,
                      radius_m: float, n_points: int = 16) -> List[List[float]]:
    """Approximate a circle as a polygon with n_points vertices."""
    points = []
    for i in range(n_points):
        bearing = 360.0 * i / n_points
        plat, plon = project_point(center_lat, center_lon, bearing, radius_m)
        points.append([plat, plon])
    return points


def _segments_intersect_2d(ax1: float, ay1: float, ax2: float, ay2: float,
                           bx1: float, by1: float, bx2: float, by2: float) -> bool:
    """Check if segment (a1-a2) intersects segment (b1-b2) using cross products.
    Coordinates are flat (lat, lon treated as planar — valid for small areas)."""
    dx_a = ax2 - ax1
    dy_a = ay2 - ay1
    dx_b = bx2 - bx1
    dy_b = by2 - by1

    denom = dx_a * dy_b - dy_a * dx_b
    if abs(denom) < 1e-15:
        return False  # parallel

    t = ((bx1 - ax1) * dy_b - (by1 - ay1) * dx_b) / denom
    u = ((bx1 - ax1) * dy_a - (by1 - ay1) * dx_a) / denom

    return 0.0 < t < 1.0 and 0.0 < u < 1.0


def segment_intersects_polygon(p1_lat: float, p1_lon: float,
                               p2_lat: float, p2_lon: float,
                               polygon: List[List[float]]) -> bool:
    """Check if segment p1-p2 intersects any edge of polygon or passes through it."""
    # If either endpoint is inside the polygon, the segment intersects
    if point_in_polygon(p1_lat, p1_lon, polygon) or point_in_polygon(p2_lat, p2_lon, polygon):
        return True
    n = len(polygon)
    if n < 3:
        return False
    for i in range(n):
        j = (i + 1) % n
        if _segments_intersect_2d(
            p1_lat, p1_lon, p2_lat, p2_lon,
            polygon[i][0], polygon[i][1], polygon[j][0], polygon[j][1]
        ):
            return True
    return False


def is_visible(p1_lat: float, p1_lon: float, p2_lat: float, p2_lon: float,
               obstacles: List[List[List[float]]]) -> bool:
    """Check if two points can see each other (no obstacle polygon blocks the line of sight)."""
    for poly in obstacles:
        if segment_intersects_polygon(p1_lat, p1_lon, p2_lat, p2_lon, poly):
            return False
    return True


def is_visible_vg(p1_lat: float, p1_lon: float, p2_lat: float, p2_lon: float,
                  obstacles: List[List[List[float]]]) -> bool:
    """Visibility check for visibility graph — handles polygon vertex endpoints.
    Unlike is_visible, does NOT check endpoint containment (vertices lie on
    polygon boundaries and would be incorrectly classified as 'inside').
    Checks edge crossings + perpendicular interior test to distinguish
    polygon-edge segments from polygon-interior segments."""
    mid_lat = (p1_lat + p2_lat) / 2
    mid_lon = (p1_lon + p2_lon) / 2

    # Perpendicular offset for interior detection (~1cm)
    dlat = p2_lat - p1_lat
    dlon = p2_lon - p1_lon
    norm = math.sqrt(dlat * dlat + dlon * dlon)
    eps = 1e-7
    if norm > 1e-15:
        perp_lat = -dlon / norm * eps
        perp_lon = dlat / norm * eps
    else:
        perp_lat, perp_lon = eps, 0.0

    cl1, cn1 = mid_lat + perp_lat, mid_lon + perp_lon
    cl2, cn2 = mid_lat - perp_lat, mid_lon - perp_lon

    for poly in obstacles:
        n = len(poly)
        if n < 3:
            continue
        # Check edge crossings (strict inequality — ignores shared endpoints)
        for i in range(n):
            j = (i + 1) % n
            if _segments_intersect_2d(
                p1_lat, p1_lon, p2_lat, p2_lon,
                poly[i][0], poly[i][1], poly[j][0], poly[j][1]
            ):
                return False
        # Both perpendicular sides inside → segment cuts through polygon interior
        # (a segment along a polygon edge has only ONE side inside)
        if point_in_polygon(cl1, cn1, poly) and point_in_polygon(cl2, cn2, poly):
            return False
    return True


def polygon_buffer(polygon: List[List[float]], buffer_m: float) -> List[List[float]]:
    """Expand polygon outward by buffer_m meters from boundary.
    Each edge is offset outward along its normal, vertices placed at
    intersections of adjacent offset edges (miter join).
    Uses flat-earth approximation for speed (accurate within ~10km)."""
    if not polygon or buffer_m <= 0:
        return list(polygon) if polygon else []

    n = len(polygon)
    if n < 3:
        return list(polygon)

    cx = sum(p[0] for p in polygon) / n
    cy = sum(p[1] for p in polygon) / n
    lat_scale = 111320.0
    lon_scale = 111320.0 * math.cos(math.radians(cx))

    # Convert to flat meters
    flat = [((p[0] - cx) * lat_scale, (p[1] - cy) * lon_scale) for p in polygon]

    # Signed area to determine winding (CCW → positive)
    area2 = 0.0
    for i in range(n):
        j = (i + 1) % n
        area2 += flat[i][0] * flat[j][1] - flat[j][0] * flat[i][1]
    sign = 1.0 if area2 > 0 else -1.0

    result = []
    for i in range(n):
        prev_i = (i - 1) % n
        next_i = (i + 1) % n

        # Incoming edge: prev → i
        dx1 = flat[i][0] - flat[prev_i][0]
        dy1 = flat[i][1] - flat[prev_i][1]
        L1 = math.sqrt(dx1 * dx1 + dy1 * dy1)

        # Outgoing edge: i → next
        dx2 = flat[next_i][0] - flat[i][0]
        dy2 = flat[next_i][1] - flat[i][1]
        L2 = math.sqrt(dx2 * dx2 + dy2 * dy2)

        if L1 < 0.01 or L2 < 0.01:
            result.append([polygon[i][0], polygon[i][1]])
            continue

        # Outward normals (for CCW: (dy, -dx); flip sign for CW)
        nx1, ny1 = sign * dy1 / L1, sign * (-dx1) / L1
        nx2, ny2 = sign * dy2 / L2, sign * (-dx2) / L2

        # Bisector direction
        bx = nx1 + nx2
        by = ny1 + ny2
        blen = math.sqrt(bx * bx + by * by)
        if blen < 1e-10:
            bx, by = nx1, ny1
        else:
            bx /= blen
            by /= blen

        # Offset along bisector: buffer / cos(half_angle_between_normals)
        cos_half = nx1 * bx + ny1 * by
        cos_half = max(cos_half, 0.25)  # cap at ~75° to prevent spikes
        offset = buffer_m / cos_half

        new_x = flat[i][0] + bx * offset
        new_y = flat[i][1] + by * offset

        new_lat = cx + new_x / lat_scale
        new_lon = cy + new_y / lon_scale
        result.append([new_lat, new_lon])

    return result


# ──────────────────── Batch (numpy) geometry ────────────────────


def batch_segments_intersect(segs_a: np.ndarray, segs_b: np.ndarray) -> np.ndarray:
    """Vectorized intersection test between two sets of line segments.

    Args:
        segs_a: (P, 4) array — [lat1, lon1, lat2, lon2] per row
        segs_b: (E, 4) array — [lat1, lon1, lat2, lon2] per row

    Returns:
        (P, E) bool array — result[i,j] is True when seg_a[i] crosses seg_b[j].
    """
    ax1, ay1 = segs_a[:, 0], segs_a[:, 1]
    ax2, ay2 = segs_a[:, 2], segs_a[:, 3]
    bx1, by1 = segs_b[:, 0], segs_b[:, 1]
    bx2, by2 = segs_b[:, 2], segs_b[:, 3]

    dx_a = ax2 - ax1  # (P,)
    dy_a = ay2 - ay1
    dx_b = bx2 - bx1  # (E,)
    dy_b = by2 - by1

    # (P,1) × (1,E) → (P,E)
    denom = dx_a[:, None] * dy_b[None, :] - dy_a[:, None] * dx_b[None, :]

    dbx = bx1[None, :] - ax1[:, None]  # (P, E)
    dby = by1[None, :] - ay1[:, None]

    safe_denom = np.where(np.abs(denom) < 1e-15, 1.0, denom)
    t = (dbx * dy_b[None, :] - dby * dx_b[None, :]) / safe_denom
    u = (dbx * dy_a[:, None] - dby * dx_a[:, None]) / safe_denom

    return (np.abs(denom) >= 1e-15) & (t > 0) & (t < 1) & (u > 0) & (u < 1)


def batch_point_in_polygon(points: np.ndarray, polygon: np.ndarray) -> np.ndarray:
    """Ray-casting for M points against one polygon (vectorized over points).

    Args:
        points: (M, 2) array — [lat, lon] per row
        polygon: (N, 2) array — [lat, lon] per vertex

    Returns:
        (M,) bool array — True if point is inside polygon.
    """
    n = polygon.shape[0]
    if n < 3:
        return np.zeros(points.shape[0], dtype=bool)

    plat = points[:, 0]
    plon = points[:, 1]
    inside = np.zeros(len(plat), dtype=bool)

    j = n - 1
    for i in range(n):
        yi, xi = polygon[i, 0], polygon[i, 1]
        yj, xj = polygon[j, 0], polygon[j, 1]

        dy = yj - yi
        if abs(dy) < 1e-15:
            j = i
            continue

        cond = (yi > plat) != (yj > plat)
        x_intersect = (xj - xi) * (plat - yi) / dy + xi
        inside ^= cond & (plon < x_intersect)

        j = i

    return inside


def batch_haversine(lat1: np.ndarray, lon1: np.ndarray,
                    lat2: np.ndarray, lon2: np.ndarray) -> np.ndarray:
    """Vectorized haversine distance in meters."""
    lat1_r = np.radians(lat1)
    lat2_r = np.radians(lat2)
    dlat = np.radians(lat2 - lat1)
    dlon = np.radians(lon2 - lon1)
    a = np.sin(dlat / 2) ** 2 + np.cos(lat1_r) * np.cos(lat2_r) * np.sin(dlon / 2) ** 2
    c = 2 * np.arctan2(np.sqrt(a), np.sqrt(1 - a))
    return EARTH_RADIUS * c

File: cpp/src/navigation/Calculations.cpp
#include "Calculations.h"
#include <algorithm>
#include <cmath>
#include <numeric>

namespace vtol::nav {

double haversineDistance(double lat1, double lon1, double lat2, double lon2)
{
    double lat1Rad = lat1 * M_PI / 180.0;
    double lat2Rad = lat2 * M_PI / 180.0;
    double dLat = (lat2 - lat1) * M_PI / 180.0;
    double dLon = (lon2 - lon1) * M_PI / 180.0;

    double a = std::sin(dLat / 2) * std::sin(dLat / 2) +
               std::cos(lat1Rad) * std::cos(lat2Rad) *
               std::sin(dLon / 2) * std::sin(dLon / 2);
    double c = 2 * std::atan2(std::sqrt(a), std::sqrt(1 - a));
    return EARTH_RADIUS * c;
}

double bearingTo(double lat1, double lon1, double lat2, double lon2)
{
    double lat1Rad = lat1 * M_PI / 180.0;
    double lat2Rad = lat2 * M_PI / 180.0;
    double dLon = (lon2 - lon1) * M_PI / 180.0;

    double x = std::sin(dLon) * std::cos(lat2Rad);
    double y = std::cos(lat1Rad) * std::sin(lat2Rad) -
               std::sin(lat1Rad) * std::cos(lat2Rad) * std::cos(dLon);

    double bearing = std::atan2(x, y);
    return std::fmod(bearing * 180.0 / M_PI + 360.0, 360.0);
}

double crossTrackDistance(double posLat, double posLon,
                           double wp1Lat, double wp1Lon,
                           double wp2Lat, double wp2Lon)
{
    double d13 = haversineDistance(wp1Lat, wp1Lon, posLat, posLon) / EARTH_RADIUS;
    double brng13 = bearingTo(wp1Lat, wp1Lon, posLat, posLon) * M_PI / 180.0;
    double brng12 = bearingTo(wp1Lat, wp1Lon, wp2Lat, wp2Lon) * M_PI / 180.0;

    double xtd = std::asin(std::sin(d13) * std::sin(brng13 - brng12));
    return xtd * EARTH_RADIUS;
}

double alongTrackDistance(double posLat, double posLon,
                            double wp1Lat, double wp1Lon,
                            double wp2Lat, double wp2Lon)
{
    double d13 = haversineDistance(wp1Lat, wp1Lon, posLat, posLon) / EARTH_RADIUS;
    double xtd = crossTrackDistance(posLat, posLon, wp1Lat, wp1Lon, wp2Lat, wp2Lon) / EARTH_RADIUS;

    double cosD13 = std::cos(d13);
    double cosXtd = std::cos(xtd);
    if (std::abs(cosXtd) < 1e-15) return 0.0;

    double atd = std::acos(std::clamp(cosD13 / cosXtd, -1.0, 1.0));
    return atd * EARTH_RADIUS;
}

double etaSeconds(double distanceM, double groundspeedMs)
{
    if (groundspeedMs <= 0) return std::numeric_limits<double>::infinity();
    return distanceM / groundspeedMs;
}

double metersToLatOffset(double meters)
{
    return meters / 111320.0;
}

double metersToLonOffset(double meters, double latitude)
{
    return meters / (111320.0 * std::cos(latitude * M_PI / 180.0));
}

double normalizeHeading(double heading)
{
    return std::fmod(heading, 360.0);
}

double headingDifference(double h1, double h2)
{
    double diff = h2 - h1;
    while (diff > 180)  diff -= 360;
    while (diff < -180) diff += 360;
    return diff;
}

LatLon projectPoint(double lat, double lon, double bearingDeg, double distanceM)
{
    double d = distanceM / EARTH_RADIUS;
    double brg = bearingDeg * M_PI / 180.0;
    double lat1 = lat * M_PI / 180.0;
    double lon1 = lon * M_PI / 180.0;

    double lat2 = std::asin(std::sin(lat1) * std::cos(d) +
                             std::cos(lat1) * std::sin(d) * std::cos(brg));
    double lon2 = lon1 + std::atan2(std::sin(brg) * std::sin(d) * std::cos(lat1),
                                     std::cos(d) - std::sin(lat1) * std::sin(lat2));

    return {lat2 * 180.0 / M_PI, lon2 * 180.0 / M_PI};
}

// ── Zone avoidance geometry ─────────────────────────────────────────

bool pointInPolygon(double lat, double lon, const Polygon& polygon)
{
    int n = static_cast<int>(polygon.size());
    if (n < 3) return false;

    bool inside = false;
    int j = n - 1;
    for (int i = 0; i < n; ++i) {
        double yi = polygon[i].lat, xi = polygon[i].lon;
        double yj = polygon[j].lat, xj = polygon[j].lon;
        if (((yi > lat) != (yj > lat)) &&
            (lon < (xj - xi) * (lat - yi) / (yj - yi) + xi)) {
            inside = !inside;
        }
        j = i;
    }
    return inside;
}

Polygon circleToPolygon(double centerLat, double centerLon, double radiusM, int nPoints)
{
    Polygon pts;
    pts.reserve(nPoints);
    for (int i = 0; i < nPoints; ++i) {
        double bearing = 360.0 * i / nPoints;
        pts.push_back(projectPoint(centerLat, centerLon, bearing, radiusM));
    }
    return pts;
}

bool segmentsIntersect2d(double ax1, double ay1, double ax2, double ay2,
                          double bx1, double by1, double bx2, double by2)
{
    double dxA = ax2 - ax1;
    double dyA = ay2 - ay1;
    double dxB = bx2 - bx1;
    double dyB = by2 - by1;

    double denom = dxA * dyB - dyA * dxB;
    if (std::abs(denom) < 1e-15) return false;  // parallel

    double t = ((bx1 - ax1) * dyB - (by1 - ay1) * dxB) / denom;
    double u = ((bx1 - ax1) * dyA - (by1 - ay1) * dxA) / denom;

    return t > 0.0 && t < 1.0 && u > 0.0 && u < 1.0;
}

bool segmentIntersectsPolygon(double p1Lat, double p1Lon, double p2Lat, double p2Lon,
                               const Polygon& polygon)
{
    if (pointInPolygon(p1Lat, p1Lon, polygon) || pointInPolygon(p2Lat, p2Lon, polygon))
        return true;

    int n = static_cast<int>(polygon.size());
    if (n < 3) return false;

    for (int i = 0; i < n; ++i) {
        int j = (i + 1) % n;
        if (segmentsIntersect2d(
                p1Lat, p1Lon, p2Lat, p2Lon,
                polygon[i].lat, polygon[i].lon, polygon[j].lat, polygon[j].lon))
            return true;
    }
    return false;
}

bool isVisible(double p1Lat, double p1Lon, double p2Lat, double p2Lon,
               const std::vector<Polygon>& obstacles)
{
    for (const auto& poly : obstacles) {
        if (segmentIntersectsPolygon(p1Lat, p1Lon, p2Lat, p2Lon, poly))
            return false;
    }
    return true;
}

bool isVisibleVg(double p1Lat, double p1Lon, double p2Lat, double p2Lon,
                  const std::vector<Polygon>& obstacles)
{
    double midLat = (p1Lat + p2Lat) / 2;
    double midLon = (p1Lon + p2Lon) / 2;

    // Perpendicular offset for interior detection (~1cm)
    double dlat = p2Lat - p1Lat;
    double dlon = p2Lon - p1Lon;
    double norm = std::sqrt(dlat * dlat + dlon * dlon);
    constexpr double eps = 1e-7;
    double perpLat, perpLon;
    if (norm > 1e-15) {
        perpLat = -dlon / norm * eps;
        perpLon = dlat / norm * eps;
    } else {
        perpLat = eps;
        perpLon = 0.0;
    }

    double cl1 = midLat + perpLat, cn1 = midLon + perpLon;
    double cl2 = midLat - perpLat, cn2 = midLon - perpLon;

    for (const auto& poly : obstacles) {
        int n = static_cast<int>(poly.size());
        if (n < 3) continue;

        // Check edge crossings (strict inequality — ignores shared endpoints)
        for (int i = 0; i < n; ++i) {
            int j = (i + 1) % n;
            if (segmentsIntersect2d(
                    p1Lat, p1Lon, p2Lat, p2Lon,
                    poly[i].lat, poly[i].lon, poly[j].lat, poly[j].lon))
                return false;
        }
        // Both perpendicular sides inside → segment cuts through interior
        if (pointInPolygon(cl1, cn1, poly) && pointInPolygon(cl2, cn2, poly))
            return false;
    }
    return true;
}

Polygon polygonBuffer(const Polygon& polygon, double bufferM)
{
    if (polygon.empty() || bufferM <= 0)
        return polygon;

    int n = static_cast<int>(polygon.size());
    if (n < 3) return polygon;

    // Centroid
    double cx = 0, cy = 0;
    for (const auto& p : polygon) { cx += p.lat; cy += p.lon; }
    cx /= n; cy /= n;

    constexpr double latScale = 111320.0;
    double lonScale = 111320.0 * std::cos(cx * M_PI / 180.0);

    // Convert to flat meters
    struct FlatPt { double x, y; };
    std::vector<FlatPt> flat(n);
    for (int i = 0; i < n; ++i) {
        flat[i] = {(polygon[i].lat - cx) * latScale, (polygon[i].lon - cy) * lonScale};
    }

    // Signed area → winding direction
    double area2 = 0.0;
    for (int i = 0; i < n; ++i) {
        int j = (i + 1) % n;
        area2 += flat[i].x * flat[j].y - flat[j].x * flat[i].y;
    }
    double sign = area2 > 0 ? 1.0 : -1.0;

    Polygon result;
    result.reserve(n);

    for (int i = 0; i < n; ++i) {
        int prevI = (i - 1 + n) % n;
        int nextI = (i + 1) % n;

        double dx1 = flat[i].x - flat[prevI].x;
        double dy1 = flat[i].y - flat[prevI].y;
        double L1 = std::sqrt(dx1 * dx1 + dy1 * dy1);

        double dx2 = flat[nextI].x - flat[i].x;
        double dy2 = flat[nextI].y - flat[i].y;
        double L2 = std::sqrt(dx2 * dx2 + dy2 * dy2);

        if (L1 < 0.01 || L2 < 0.01) {
            result.push_back(polygon[i]);
            continue;
        }

        // Outward normals
        double nx1 = sign * dy1 / L1, ny1 = sign * (-dx1) / L1;
        double nx2 = sign * dy2 / L2, ny2 = sign * (-dx2) / L2;

        // Bisector direction
        double bx = nx1 + nx2;
        double by = ny1 + ny2;
        double blen = std::sqrt(bx * bx + by * by);
        if (blen < 1e-10) {
            bx = nx1; by = ny1;
        } else {
            bx /= blen; by /= blen;
        }

        // Offset along bisector
        double cosHalf = std::max(nx1 * bx + ny1 * by, 0.25);
        double offset = bufferM / cosHalf;

        double newX = flat[i].x + bx * offset;
        double newY = flat[i].y + by * offset;

        result.push_back({cx + newX / latScale, cy + newY / lonScale});
    }

    return result;
}

} // namespace vtol::nav
